python/lsst/daf/butler/server.py

```diff
@@ -0,0 +1,441 @@
+GLOBAL_READONLY_BUTLER = None
+
+
+def _make_global_butler() -> None:
+    global GLOBAL_READONLY_BUTLER, GLOBAL_READWRITE_BUTLER
+    if GLOBAL_READONLY_BUTLER is None:
+        GLOBAL_READONLY_BUTLER = Butler(BUTLER_ROOT, writeable=False)
+    if GLOBAL_READWRITE_BUTLER is None:
+        GLOBAL_READWRITE_BUTLER = Butler(BUTLER_ROOT, writeable=True)
+
+
+def butler_readonly_dependency() -> Butler:
+    _make_global_butler()
+    return Butler(butler=GLOBAL_READONLY_BUTLER)
+
+
+def butler_readwrite_dependency() -> Butler:
+    _make_global_butler()
+    return Butler(butler=GLOBAL_READWRITE_BUTLER)
+
+
+def unpack_dataId(butler: Butler, data_id: SerializedDataCoordinate | None) -> DataCoordinate | None:
+    """Convert the serialized dataId back to full DataCoordinate.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The butler to use for registry and universe.
+    data_id : `SerializedDatacoordinate` or `None`
+        The serialized form.
+
+    Returns
+    -------
+    dataId : `DataCoordinate` or `None`
+        The DataId usable by registry.
+    """
+    if data_id is None:
+        return None
+    return DataCoordinate.from_simple(data_id, registry=butler.registry)
+
+
+@app.get("/butler/")
+def read_root() -> str:
+    return "Welcome to Excalibur... aka your Butler Server"
+
+
+@app.get("/butler/butler.json", response_model=dict[str, Any])
+def read_server_config() -> Mapping:
+    """Return the butler configuration that the client should use."""
+    config_str = f"""
+datastore:
+    root: {BUTLER_ROOT}
+registry:
+    cls: lsst.daf.butler.registries.remote.RemoteRegistry
+    db: <butlerRoot>
+"""
+    config = Config.fromString(config_str, format="yaml")
+    return config
+
+
+@app.get("/butler/universe", response_model=dict[str, Any])
+def get_dimension_universe(butler: Butler = Depends(butler_readonly_dependency)) -> DimensionConfig:
+    """Allow remote client to get dimensions definition."""
+    return butler.registry.dimensions.dimensionConfig
+
+
+@app.get("/butler/v1/uri/{id}", response_model=str)
+def get_uri(id: DatasetId, butler: Butler = Depends(butler_readonly_dependency)) -> str:
+    """Return a single URI of non-disassembled dataset."""
+    ref = butler.registry.getDataset(id)
+    if not ref:
+        raise HTTPException(status_code=404, detail=f"Dataset with id {id} does not exist.")
+
+    uri = butler.datastore.getURI(ref)
+
+    # In reality would have to convert this to a signed URL
+    return str(uri)
+
+
+@app.put("/butler/v1/registry/refresh")
+def refresh(butler: Butler = Depends(butler_readonly_dependency)) -> None:
+    # Unclear whether this should exist. Which butler is really being
+    # refreshed? How do we know the server we are refreshing is used later?
+    # For testing at the moment it is important if a test adds a dataset type
+    # directly in the server since the test client will not see it.
+    butler.registry.refresh()
```

Comment:
Probably, but I'm having an existential crisis over the refreshing in general. As you say in the overall comment, caching the dataset types and collections in the long-lived server butlers is going to cause a lot of problems since we never really know when something has changed in the database backend that would need a refresh. Globally disabling the cache wold be safest but might be a performance disaster when 10,000 people connect. cc/ @TallJimbo